tests/tests.py

```diff
@@ -185,3 +189,165 @@ def test_each_context(self):
+        """
+        model = CustomAdminModel.objects.create(name="aaaa")
+        # filter model list by some parameter
+        query = "name__startswith=a"
+        changelist_url = get_changelist_url(CustomAdminModel) + "?" + query
+        # choose mass change action
+        response = self.client.post(changelist_url,
+                                    {"action": "mass_change_selected",
+                                     "_selected_action": model.pk})
+        self.assertEqual(response.status_code, 302)
+        mass_change_url = response.get("Location")
+        # filters add to redirect url
+        self.assertEqual(parse.parse_qs(parse.urlparse(mass_change_url).query),
+                         {"_changelist_filters": [query]})
+        # save mass change form with preserved filters
+        response = self.client.post(mass_change_url,
+                                    {"_mass_change": "name", "name": "new name"})
+        # we are redirected to changelist with filters
+        self.assertRedirects(response, changelist_url)
+
+    def test_invalid_form(self):
+        """ Save nothing if some forms are invalid
+        """
+        models = [CustomAdminModel.objects.create(name="model {}".format(i))
+                  for i in range(0, 3)]
+
+        response = self.client.post(async_get_massadmin_url(models, self.client.session),
+                                    {"_mass_change": "name",
+                                     "name": "invalid {}".format(models[-1].pk)})
+        self.assertEqual(response.status_code, 200)
+        self.assertContains(response, 'errornote')
+        new_names = CustomAdminModel.objects.order_by("pk").values_list("name", flat=True)
+        # all models stay the same
+        self.assertEqual(list(new_names), [m.name for m in models])
+
+    def test_get_fieldsets(self):
+        """ Use the admin_obj's get_fieldsets method if it is defined
+        """
+        names = [
+            {
+                "first_name": "first {}".format(i),
+                "middle_name": "middle {}".format(i),
+                "last_name": "last {}".format(i)
+            }
+            for i in range(3)
+        ]
+        models = [FieldsetsAdminModel.objects.create(**n) for n in names]
+        response = self.client.get(async_get_massadmin_url(models, self.client.session))
+        self.assertEqual(response.status_code, 200)
+
+        expected_fieldset_group_1 = """
+        <h2 class="grp-collapse-handler">First part of name Mass Edit</h2>
+        """
+        expected_fieldset_group_2 = """
+        <h2 class="grp-collapse-handler">Second part of name Mass Edit</h2>
+        """
+
+        self.assertContains(response, expected_fieldset_group_1, html=True)
+        self.assertContains(response, expected_fieldset_group_2, html=True)
+
+
+class AsyncCustomizationTestCase(TestCase):
+    """ ImprovedMassAdmin has all customized options from related ModelAdmin
+    """
```

Comment:
```suggestion
    """ImprovedMassAdmin has all customized options from related ModelAdmin"""
```
and at other places.